### PR TITLE
[SERVICE-569] Implement Toast UX

### PR DESCRIPTION
**Pull request by ngunday**
- Seperates layouting logic from Toasts and ToastManager, implements Layouter module.
- Implements new animations for the Toast UX
- Implements adjustable anchor for toasts

### CODE REVIEW

**Comment by finos-admin** [comment]
Thank you for your pull request and welcome to our community! We require contributors to sign a [Contributor License Agreement](https://finosfoundation.atlassian.net/wiki/spaces/FINOS/pages/75530375/Legal+Requirements#LegalRequirements-ContributorLicenseAgreement) and we don't seem to have CLAs on file for these contributors to the Pull Request: (@ngunday). In order for your PR to be reviewed and merged, please follow the directions at the link above.

Project team: please **do not merge this Pull Request** until Foundation staff have confirmed that a CLA is in place for the new contributor(s) listed above.

If there are any questions, please don't hesitate to [get in touch with our Infrastructure Support team](https://finosfoundation.atlassian.net/wiki/spaces/FINOS/pages/77955298/Getting+in+Touch#GettinginTouch-InfrastructureSupport). /CC @finos-admin

**Comment by finos-admin** [comment]
Thank you for your pull request and welcome to our community! We require contributors to sign a [Contributor License Agreement](https://finosfoundation.atlassian.net/wiki/spaces/FINOS/pages/75530375/Legal+Requirements#LegalRequirements-ContributorLicenseAgreement) and we don't seem to have CLAs on file for these contributors to the Pull Request: (@ngunday). In order for your PR to be reviewed and merged, please follow the directions at the link above.

Project team: please **do not merge this Pull Request** until Foundation staff have confirmed that a CLA is in place for the new contributor(s) listed above.

If there are any questions, please don't hesitate to [get in touch with our Infrastructure Support team](https://finosfoundation.atlassian.net/wiki/spaces/FINOS/pages/77955298/Getting+in+Touch#GettinginTouch-InfrastructureSupport). /CC @finos-admin

**Comment by finos-admin** [comment]
Thank you for your pull request and welcome to our community! We require contributors to sign a [Contributor License Agreement](https://finosfoundation.atlassian.net/wiki/spaces/FINOS/pages/75530375/Legal+Requirements#LegalRequirements-ContributorLicenseAgreement) and we don't seem to have CLAs on file for these contributors to the Pull Request: (@ngunday). In order for your PR to be reviewed and merged, please follow the directions at the link above.

Project team: please **do not merge this Pull Request** until Foundation staff have confirmed that a CLA is in place for the new contributor(s) listed above.

If there are any questions, please don't hesitate to [get in touch with our Infrastructure Support team](https://finosfoundation.atlassian.net/wiki/spaces/FINOS/pages/77955298/Getting+in+Touch#GettinginTouch-InfrastructureSupport). /CC @finos-admin

File: src/provider/controller/Layouter.ts [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        const origin: PointTopLeft = {top: this._availableRect.bottom / 2, left: this._availableRect.right / 2};
+        return {
+            top: origin.top + origin.top * this.anchor.top + this.margin,
+            left: origin.left + origin.left * this.anchor.left
+        };
+    }
+
+    /**
+     * Layout a given stack of Layoutable items
+     * @param item {LayoutItem} Target layoutable item stack
```
[the remark this answers is not in the document]
Comment (ngunday) [reply src/provider/controller/Layouter.ts]: done

File: src/provider/controller/Layouter.ts [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        fin.System.addListener('monitor-info-changed', (async (event: MonitorEvent<string, string>) => {
+            const monitorInfo = await fin.System.getMonitorInfo();
+            this._availableRect = monitorInfo.primaryMonitor.availableRect;
+            Layouter.eventEmitter.emit(LayoutEvent.LAYOUT_REQUIRED);
+        }));
+    }
+
+    /**
+     * Anchor point of the layout. clamped between -1 (for minimum) and 1 (for maximum) for each dimension.
+     * @return {PointTopLeft} anchor position
```
[the remark this answers is not in the document]
Comment (ngunday) [reply src/provider/controller/Layouter.ts]: done

File: src/provider/controller/Layouter.ts [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    animate(transitions: Transition, options: TransitionOptions): Promise<void>;
+    setTransform(transform: Bounds): Promise<void>;
+    readonly dimensions: Promise<WindowDimensions>;
+    position: PointTopLeft;
+}
+
+export type WindowDimensions = {height: number, width: number};
+
+export enum LayoutEvent {
+    LAYOUT_REQUIRED = 'layoutRequired'
```
[the remark this answers is not in the document]
Comment (ngunday) [reply src/provider/controller/Layouter.ts]: done

File: src/provider/controller/ToastManager.ts [hunk 4]
```diff
@@ -1,29 +1,22 @@
...
 import {Action, RootAction} from '../store/Actions';
 import {Store} from '../store/Store';
-
-export type WindowDimensions = {height: number, width: number};
+import {LayoutEvent, Layouter} from './Layouter';
 
 @injectable()
 export class ToastManager {
+    @inject(Inject.LAYOUTER)
+    private _layouter!: Layouter;
```
[the remark this answers is not in the document]
Comment (ngunday) [reply src/provider/controller/ToastManager.ts]: By design, the Layouter is intended act to be an abstract service completely agnostic to Toasts or the ToastManager and provide an interface for any module that requires a layout support. I believe in a seperation of concerns point of view, there's no real reason for ToastManager as a client to be concerned about the construction of this service.

**Comment by finos-admin** [comment]
Thank you for your pull request and welcome to our community! We require contributors to sign a [Contributor License Agreement](https://finosfoundation.atlassian.net/wiki/spaces/FINOS/pages/75530375/Legal+Requirements#LegalRequirements-ContributorLicenseAgreement) and we don't seem to have CLAs on file for these contributors to the Pull Request: (@ngunday). In order for your PR to be reviewed and merged, please follow the directions at the link above.

Project team: please **do not merge this Pull Request** until Foundation staff have confirmed that a CLA is in place for the new contributor(s) listed above.

If there are any questions, please don't hesitate to [get in touch with our Infrastructure Support team](https://finosfoundation.atlassian.net/wiki/spaces/FINOS/pages/77955298/Getting+in+Touch#GettinginTouch-InfrastructureSupport). /CC @finos-admin

File: src/provider/controller/Layouter.ts [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+export class Layouter {
+    private static INTERNAL_CONFIG: Readonly<LayouterConfig> = {
+        spacing: 10,
+        anchor: {top: 1, left: 1},
+        animationTime: 100
+    };
+
+    private _availableRect!: Required<Rect>;
+
+    public static eventEmitter: EventEmitter = new EventEmitter();
```
Comment (tomrobertsOF) [inline src/provider/controller/Layouter.ts]: Are we sure an eventEmitter is the right choice here? We've been using `Signals` for most similar things in this and the other services.

File: src/provider/controller/ToastManager.ts [hunk 6]
```diff
@@ -91,6 +84,14 @@ export class ToastManager {
         });
     }
 
+    /**
+     * Signal callback for layout required event. This signal is emitted by Layouter in events like
+     * monitor info is changed, etc.
+     */
+    private onLayoutRequired(): void {
```
[the remark this answers is not in the document]
Comment (ngunday) [reply src/provider/controller/ToastManager.ts]: Layouter does not own or keeps track of the stack, it is a data operator for any provided stack.

File: src/provider/model/Toast.ts [hunk 7]
```diff
@@ -42,27 +46,17 @@ interface Options {
...
     PAUSE = 'pause',
     UNPAUSE = 'unpause',
     CLOSED = 'closed'
 }
 
-
-export class Toast {
-    public static DIRECTION: Readonly<[number, number]> = [-1, 1];
+export class Toast implements LayoutItem {
     public static eventEmitter: EventEmitter = new EventEmitter();
```
[the remark this answers is not in the document]
Comment (ngunday) [reply src/provider/model/Toast.ts]: done

File: src/provider/controller/ToastManager.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (pjbroadbent) [reply src/provider/controller/ToastManager.ts]: Ahh, my bad - missed the arg.

**Pull request merged**